### ISSUE TITLE
MemDFJK not using in-core algorithm

### ISSUE THREAD

**Issue by susilehtola** [psi4/psi4#1666]
Excerpt of a density-fitted calculation
```
  ==> Integral Setup <==

  DFHelper Memory: AOs need 5.003 GiB; user supplied 19.539 GiB. Using in-core AOs.

  ==> MemDFJK: Density-Fitted J/K Matrices <==

    J tasked:                   Yes
    K tasked:                    No
    wK tasked:                   No
    OpenMP threads:               4
    Memory [MiB]:             20008
    Algorithm:                 Disk
    Schwarz Cutoff:           1E-12
    Mask sparsity (%):      35.4523
    Fitting Condition:        1E-10
```
That is, there's two bugs here: the code is not using in-core AOs, and the printout doesn't match what the code is actually doing.

**Comment by loriab** [comment]
Would you post the input that triggered this, please? I had thought we had the disk/mem routing printing right.

**Comment by susilehtola** [comment]
I think this was the one
```
molecule {
pubchem:91749189
}

memory 40 gb
set basis pcseg-1
set guess sap
energy('pbe')
```

**Comment by susilehtola** [comment]
So for clarification: it is using MemDFJK, but it's not using in-core AOs in MemDFJK (whatever that means).

**Comment by obrien951** [comment]
Hi Dr. Lehtola,

Would you be able to post the output file as well? I was hoping to know exactly what code you're using because there have been a lot of revisions to JK algorithm selection based on memory lately. This would help me look at the code block that selected your algorithm and possibly help with your first bug.

This bug usually shows up when the user selects an "SCF_TYPE" and then doesn't supply an amount of memory that warrants Disk/core.

The second bug isn't so much a bug as a troublesome name. MemDFJK is a JK build package that is optimized to work in the case that there is enough core memory to hold the 3-index integral tensors, and DiskDFJK is a JK build package that we found performs better in the case where these tensors have to be read from disk. They were named after the cases in which they worked best for us not for their capabilities or limitations. 

That being said, each of these packages has machinery to handle both in-core and disk-based integral handling. Psi4 chooses the package based on the value of the  memory_ variable, and each package should then use the algorithm ("Disk" or "Core") that best fits the case.

**Comment by obrien951** [comment]
What's strange is that DFHelper thinks it has 19 gib, it only needs 5 gib, but it insists on using its disk machinery. It could just be a printing error where the code is printing out the wrong message. The calculation should give you a "timer.dat" file too. can you post that as well? That one has timers around the disk writes, so we can see if they're actually happening.

**Comment by susilehtola** [comment]
Log file attached.
[memdfjk.txt](https://github.com/psi4/psi4/files/3302510/memdfjk.txt)

The issue is simple, however, and the fix is in #1667 as I posted already above.

**Comment by dgasmith** [comment]
Ok, the memory makes sense if confusing. 37.25 GiB supplied, 0.75 SCF safety factor so 28 GiB passed to SCF. The collection took 8 GiB so 20 GiB given to JK (all extra is fed there).

Still a bit confused why Disk is used though and why #1667 is needed.

**Comment by susilehtola** [comment]
I put in a debug statement in the `AO_core()` routine
```
    // a fraction of memory to use, do we want it as an option?                                                                                                                               
    if (memory_ < required_core_size_) {
      outfile->Printf("memory = %lu < required_core_size = %lu, turning core off\n",(long unsigned) memory_,(long unsigned) required_core_size_);
      AO_core_ = false;
    }
```
and I see the following in the log file
```
  ==> Primary Basis <==

  Basis Set: PCSEG-0
    Blend: PCSEG-0
    Number of shells: 313
    Number of basis function: 489
    Number of Cartesian functions: 489
    Spherical Harmonics?: true
    Max angular momentum: 1

   => Loading Basis Set <=

    Name: (PCSEG-0 AUX)
    Role: JKFIT
    Keyword: DF_BASIS_SCF
    atoms 1-2   entry SI         line   742 file /home/work/psi4/install/share/psi4/basis/def2-sv_p_-jkfit.gbs 
    atoms 3     entry F          line   383 file /home/work/psi4/install/share/psi4/basis/def2-sv_p_-jkfit.gbs 
    atoms 4     entry O          line   323 file /home/work/psi4/install/share/psi4/basis/def2-sv_p_-jkfit.gbs 
    atoms 5-8   entry N          line   263 file /home/work/psi4/install/share/psi4/basis/def2-sv_p_-jkfit.gbs 
    atoms 9-43  entry C          line   203 file /home/work/psi4/install/share/psi4/basis/def2-sv_p_-jkfit.gbs 
    atoms 44-90 entry H          line    23 file /home/work/psi4/install/share/psi4/basis/def2-sv_p_-jkfit.gbs 

memory = 256000000 < required_core_size = 671520055, turning core off
memory = 256000000 < required_core_size = 671520055, turning core off
  ==> Pre-Iterations <==

   -------------------------------------------------------
    Irrep   Nso     Nmo     Nalpha   Nbeta   Ndocc  Nsocc
   -------------------------------------------------------
     A        489     489       0       0       0       0
   -------------------------------------------------------
    Total     489     489     165     165     165       0
   -------------------------------------------------------

  ==> Integral Setup <==

  DFHelper Memory: AOs need 5.003 GiB; user supplied 27.940 GiB. Turning off in-core AOs.
```

That is, the routine gets called twice in the basis set up phase with the default memory 256 MB. This is why the initialization is screwed up. Someone else's job to fix that ;)

**Comment by schiebermc** [comment]
ah! thanks for that. it looks like it's called in `get_core_size();`, this issue is probably resolved if 229 assigns in both outcomes.

**Comment by susilehtola** [comment]
> ah! thanks for that. it looks like it's called in `get_core_size();`, this issue is probably resolved if 229 assigns in both outcomes.

Um, @schiebermc, you mean you are fine with the patch in #1667  as-is..?

**Comment by schiebermc** [comment]
> > ah! thanks for that. it looks like it's called in `get_core_size();`, this issue is probably resolved if 229 assigns in both outcomes.
> 
> Um, @schiebermc, you mean you are fine with the patch in #1667 as-is..?

Yes. Although others who are actively moving the code around might want to consider if there are any repercussions to calling `AO_core()` with the default `memory_`.  I do not think so. 

`get_core_size()` is just used when considering which object to build, MemDFJK or DiskDFJK in: https://github.com/psi4/psi4/blob/a6305a1b251d0812ecc93438bc907a91964b0ca9/psi4/src/psi4/libfock/jk.cc#L172

As you pointed out, this call had the unintended consequence of reassigning `AO_core_`. Another solution could be to make sure `memory_` gets set (so the default is not used) before:
https://github.com/psi4/psi4/blob/a6305a1b251d0812ecc93438bc907a91964b0ca9/psi4/src/psi4/libfock/MemDFJK.cc#L68

Unfortunately, this might have resulted in the wrong code (disk based MemDFJK) being used where `required_core_size_` \in [256M, memory_].

**Comment by andysim** [comment]
Fixed in #1667
